File: src/test/java/com/flexpoker/signup/command/aggregate/SignUpUserTest.java
package com.flexpoker.signup.command.aggregate;

import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.util.UUID;

import org.junit.jupiter.api.Test;
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;

import com.flexpoker.exception.FlexPokerException;
import com.flexpoker.signup.SignUpUser;

public class SignUpUserTest {

    private static final String VALID_EMAIL_ADDRESS = "unique@test.com";

    private static final String VALID_USERNAME = "unique";

    private static final String VALID_PASSWORD = "123456";

    private static final String VALID_ENCRYPTED_PASSWORD = new BCryptPasswordEncoder().encode(VALID_PASSWORD);

    private static final UUID VALID_AGGREGATE_ID = UUID.randomUUID();

    private static final UUID VALID_SIGN_UP_CODE = UUID.randomUUID();

    @Test
    void testConfirmSignedUpUserSucceedsEvent() {
        var signUpUser = new SignUpUser(VALID_AGGREGATE_ID, VALID_SIGN_UP_CODE, VALID_EMAIL_ADDRESS,
                VALID_USERNAME, VALID_ENCRYPTED_PASSWORD);
        signUpUser.confirmSignedUpUser(VALID_USERNAME, VALID_SIGN_UP_CODE);
        assertTrue(signUpUser.isConfirmed());
    }

    @Test
    void testConfirmSignedUpUserFailsBadUsername() {
        var signUpUser = new SignUpUser(VALID_AGGREGATE_ID, VALID_SIGN_UP_CODE, VALID_EMAIL_ADDRESS,
                VALID_USERNAME, VALID_ENCRYPTED_PASSWORD);
        assertThrows(FlexPokerException.class,
                () -> signUpUser.confirmSignedUpUser("notequalusername", VALID_SIGN_UP_CODE));
    }

    @Test
    void testConfirmSignedUpUserFailsBadSignUpCode() {
        var signUpUser = new SignUpUser(VALID_AGGREGATE_ID, VALID_SIGN_UP_CODE, VALID_EMAIL_ADDRESS,
                VALID_USERNAME, VALID_ENCRYPTED_PASSWORD);
        assertThrows(FlexPokerException.class,
                () -> signUpUser.confirmSignedUpUser(VALID_USERNAME, UUID.randomUUID()));
    }

    @Test
    void testConfirmSignedUpUserFailedWhenEventAlreadyApplied() {
        var signUpUser = new SignUpUser(VALID_AGGREGATE_ID, VALID_SIGN_UP_CODE, VALID_EMAIL_ADDRESS,
                VALID_USERNAME, VALID_ENCRYPTED_PASSWORD);
        signUpUser.confirmSignedUpUser(VALID_USERNAME, VALID_SIGN_UP_CODE);
        assertThrows(IllegalStateException.class,
                () -> signUpUser.confirmSignedUpUser(VALID_USERNAME, VALID_SIGN_UP_CODE));
    }

}

File: src/test/kotlin/com/flexpoker/signup/command/aggregate/SignUpUserTest.kt
package com.flexpoker.signup.command.aggregate

import com.flexpoker.exception.FlexPokerException
import com.flexpoker.signup.SignUpUser
import com.flexpoker.util.encodePassword
import org.junit.jupiter.api.Assertions.assertThrows
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test
import java.util.UUID

class SignUpUserTest {

    companion object {
        private const val VALID_EMAIL_ADDRESS = "unique@test.com"
        private const val VALID_USERNAME = "unique"
        private const val VALID_PASSWORD = "123456"
        private val VALID_ENCRYPTED_PASSWORD = encodePassword(VALID_PASSWORD)
        private val VALID_AGGREGATE_ID = UUID.randomUUID()
        private val VALID_SIGN_UP_CODE = UUID.randomUUID()
    }

    @Test
    fun testConfirmSignedUpUserSucceedsEvent() {
        val signUpUser = SignUpUser(
            VALID_AGGREGATE_ID, VALID_SIGN_UP_CODE, VALID_EMAIL_ADDRESS,
            VALID_USERNAME, VALID_ENCRYPTED_PASSWORD
        )
        signUpUser.confirmSignedUpUser(VALID_USERNAME, VALID_SIGN_UP_CODE)
        assertTrue(signUpUser.isConfirmed)
    }

    @Test
    fun testConfirmSignedUpUserFailsBadUsername() {
        val signUpUser = SignUpUser(
            VALID_AGGREGATE_ID, VALID_SIGN_UP_CODE, VALID_EMAIL_ADDRESS,
            VALID_USERNAME, VALID_ENCRYPTED_PASSWORD
        )
        assertThrows(FlexPokerException::class.java) {
            signUpUser.confirmSignedUpUser("notequalusername", VALID_SIGN_UP_CODE)
        }
    }

    @Test
    fun testConfirmSignedUpUserFailsBadSignUpCode() {
        val signUpUser = SignUpUser(
            VALID_AGGREGATE_ID, VALID_SIGN_UP_CODE, VALID_EMAIL_ADDRESS,
            VALID_USERNAME, VALID_ENCRYPTED_PASSWORD
        )
        assertThrows(FlexPokerException::class.java) {
            signUpUser.confirmSignedUpUser(VALID_USERNAME, UUID.randomUUID())
        }
    }

    @Test
    fun testConfirmSignedUpUserFailedWhenEventAlreadyApplied() {
        val signUpUser = SignUpUser(
            VALID_AGGREGATE_ID, VALID_SIGN_UP_CODE, VALID_EMAIL_ADDRESS,
            VALID_USERNAME, VALID_ENCRYPTED_PASSWORD
        )
        signUpUser.confirmSignedUpUser(VALID_USERNAME, VALID_SIGN_UP_CODE)
        assertThrows(IllegalStateException::class.java) {
            signUpUser.confirmSignedUpUser(VALID_USERNAME, VALID_SIGN_UP_CODE)
        }
    }

}